src/Composer/Repository/ComposerRepository.php

```diff
@@ -562,7 +562,11 @@ protected function loadRootServerFile()
     protected function canonicalizeUrl($url)
     {
         if ('/' === $url[0]) {
-            return preg_replace('{(https?://[^/]+).*}i', '$1' . $url, $this->url);
+            if (preg_match('{[^:]+://[^/]*}', $this->url, $matches)) {
```

Comment:
what do you think of

```suggestion
            if (preg_match('{^[^:]+://[^/]+$}', $this->url, $matches)) {
```

Reply:
I like the `^` anchor, it's stricter and faster. Added :)

The `+` quantifier does not pass the `file://` test. The hostname can be omitted and then the pattern doesn't match anymore. (file URI syntax: [RFC 8089](https://tools.ietf.org/html/rfc8089#section-2))

The `$` anchor also doesn't pass because `$this->url` can contain slashes.